src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py

```diff
@@ -15690,8 +15690,120 @@ def dt_round(
+            ErrorMessage.not_implemented(
+                "Snowpark pandas 'Series.dt.round' method doesn't yet support 'nonexistent' parameter"
+            )
+        internal_frame = self._modin_frame
+
+        slice_length, slice_unit = rule_to_snowflake_width_and_slice_unit(
+            rule=freq  # type:  ignore[arg-type]
+        )
+
+        if slice_unit not in SUPPORTED_DT_FLOOR_CEIL_FREQS or slice_unit == "second":
+            ErrorMessage.not_implemented(
+                f"Snowpark pandas 'Series.dt.round' method doesn't support setting 'freq' parameter to '{freq}'"
+            )
+
+        # We need to implement the algorithm for rounding half to even whenever
+        # the date value is at half point of the slice:
+        # https://en.wikipedia.org/wiki/Rounding#Rounding_half_to_even
+
+        # First, we need to calculate the length of half a slice.
+        # This is straightforward if the lenght is already even.
+        # If not, we then need to first downlevel the freq to a
+        # lower granularity to ensure that it is even.
+
+        def down_level_freq(slice_length: int, slice_unit: str) -> tuple[int, str]:
+            if slice_unit == "minute":
+                slice_length *= 60
+                slice_unit = "second"
+            elif slice_unit == "hour":
+                slice_length *= 60
+                slice_unit = "minute"
+            elif slice_unit == "day":
+                slice_length *= 24
+                slice_unit = "hour"
+            else:
+                f"Snowpark pandas 'Series.dt.round' method doesn't support setting 'freq' parameter with '{slice_unit}' unit"
+            return slice_length, slice_unit
+
+        if slice_length % 2 == 1:
+            slice_length, slice_unit = down_level_freq(slice_length, slice_unit)
+        half_slice_length = int(slice_length / 2)
+        base_column = col(internal_frame.data_column_snowflake_quoted_identifiers[0])
+
+        # Second, we determine whether floor represents an even number of slices.
+        # To do so, we must divide the number of epoch seconds in it over the number
+        # of epoch seconds in one slice. This way, we can get the number of slices.
+
+        floor_column = builtin("time_slice")(
+            base_column, slice_length, slice_unit, "START"
+        )
+        ceil_column = builtin("time_slice")(
+            base_column, slice_length, slice_unit, "END"
+        )
+
+        def slice_length_when_unit_is_second(slice_length: int, slice_unit: str) -> int:
+            while slice_unit != "second":
+                slice_length, slice_unit = down_level_freq(slice_length, slice_unit)
+            return slice_length
+
+        floor_epoch_seconds_column = builtin("extract")("epoch_second", floor_column)
+        floor_num_slices_column = cast(
+            floor_epoch_seconds_column
+            / pandas_lit(slice_length_when_unit_is_second(slice_length, slice_unit)),
+            IntegerType(),
```

Comment:
If the input is a float here, how does the rounding for cast work?

Reply:
Yes, the idea here is to do integer division. So if the result of the division is a float, then it will get truncated to an integer.